File: algorithm/wordifier.py
"""
Module provides Wordifier with main functionality:
    wordify number a given phone number
    all possible wordification of given phone number
    convert wordified number to phone number
"""
from algorithm.dictionary import Dictionary
from utils.utils import find_number_to_character_mapping, string_to_number, \
     remove_non_alpha_numeric

class Wordifier:
    """
    Main methods includes:
        - number_to_words
        - all_wordifications
        - words_to_number
    """
    def __init__(self):
        self.dictionary = Dictionary()
    
    def _is_valid_query(self, number_query):
        number_query = number_query[2:]
        if "0" in number_query or "1" in number_query:
            raise ValueError
        return True

    def all_wordifications(self, number):
        """
        Returns All wordification
        """
        query = remove_non_alpha_numeric(number)
        assert self._is_valid_query(query)
        result = []
        for i in range(2, len(query)):
            query_to_combinations = query[i:]
            all_combinations = self.find_all_combinations(query_to_combinations)
            words = self.check_words_against_dictionary(all_combinations)
            for word in words:
                ans = query[:i] + word + query[i+len(word):]
                result.append(ans)
        return result

    @staticmethod
    def words_to_number(query_string):
        """
        converts wordified phone number into numeric phone number
        """
        return string_to_number(query_string)

    def number_to_words(self, number):
        """
        Return word with biggest match
        """
        query = remove_non_alpha_numeric(number)
        result = ""
        last_word = ""
        assert self._is_valid_query(query)
        for i in range(2, len(query)):
            query_to_combinations = query[i:]
            all_combinations = self.find_all_combinations(query_to_combinations)
            words = self.check_words_against_dictionary(all_combinations)
            for word in words:
                ans = query[:i] + word + query[i+len(word):]
                if len(word) >= len(last_word):
                    result = ans
                    last_word = word
        return result


    def add_words_to_dictionary(self, filename):
        """
        Add words from file in the dictionary
        """
        self.dictionary.add_to_dictionary_from_txt(filename)

    def add_single_word_to_dictionary(self, word):
        """
        Add single word in the dictionary
        """
        self.dictionary.add_single_word(word)

    def check_words_against_dictionary(self, all_combinations):
        """
        Return all the combinations which are present in a Number
        """
        found_words = []
        for combination in all_combinations:
            query = combination.lower()
            is_present = self.dictionary.search_multi_words(query)
            if is_present:
                found_words.append(combination)
        return found_words

    @staticmethod
    def find_all_combinations(query_number):
        """
        From the query number, find all the combinations of words possible
        """
        all_combinations = []
        ans = find_number_to_character_mapping(query_number[0])
        curr_ans = []
        for iterator in range(1, len(query_number)):
            num = query_number[iterator]
            chars = find_number_to_character_mapping(num)
            for each_ans in ans:
                for char in chars:
                    curr_ans.append(each_ans + str(char))
                    all_combinations.append(each_ans + str(char))
            ans = curr_ans
            curr_ans = []
        return all_combinations

File: assets/js/wordifier.js
/**
 * Wordifier, ported from algorithm/wordifier.py, algorithm/dictionary.py and utils/utils.py.
 */
const NUMBER_TO_CHARACTERS = {
  '2': ['A', 'B', 'C'],
  '3': ['D', 'E', 'F'],
  '4': ['G', 'H', 'I'],
  '5': ['J', 'K', 'L'],
  '6': ['M', 'N', 'O'],
  '7': ['P', 'Q', 'R', 'S'],
  '8': ['T', 'U', 'V'],
  '9': ['W', 'X', 'Y', 'Z'],
};

const CHARACTER_TO_NUMBER = {};
for (const [digit, chars] of Object.entries(NUMBER_TO_CHARACTERS)) {
  chars.forEach((char) => { CHARACTER_TO_NUMBER[char] = digit; });
}

function findNumberToCharacterMapping(digit) {
  return NUMBER_TO_CHARACTERS[digit] || null;
}

function removeNonAlphaNumeric(input) {
  // mirrors Python's re.sub(r'\W+', '', input) — \w is [A-Za-z0-9_]
  return input.replace(/[^a-zA-Z0-9_]+/g, '');
}

function stringToNumber(text) {
  let ans = '';
  for (const char of text) {
    if (/[a-zA-Z]/.test(char)) {
      ans += CHARACTER_TO_NUMBER[char.toUpperCase()];
    } else {
      ans += char;
    }
  }
  return ans;
}

function findAllCombinations(queryNumber) {
  const allCombinations = [];
  let ans = findNumberToCharacterMapping(queryNumber[0]);
  let currAns = [];
  for (let iterator = 1; iterator < queryNumber.length; iterator++) {
    const chars = findNumberToCharacterMapping(queryNumber[iterator]);
    for (const eachAns of ans) {
      for (const char of chars) {
        currAns.push(eachAns + char);
        allCombinations.push(eachAns + char);
      }
    }
    ans = currAns;
    currAns = [];
  }
  return allCombinations;
}

class WordifierError extends Error {}

class Wordifier {
  constructor(trie) {
    this.trie = trie;
  }

  _isValidQuery(numberQuery) {
    const rest = numberQuery.slice(2);
    if (rest.includes('0') || rest.includes('1')) {
      throw new WordifierError('Digits after the first two cannot contain 0 or 1.');
    }
    return true;
  }

  checkWordsAgainstDictionary(allCombinations) {
    const foundWords = [];
    for (const combination of allCombinations) {
      if (this.trie.searchMultiWords(combination.toLowerCase())) {
        foundWords.push(combination);
      }
    }
    return foundWords;
  }

  allWordifications(number) {
    const query = removeNonAlphaNumeric(number);
    this._isValidQuery(query);
    const result = [];
    for (let i = 2; i < query.length; i++) {
      const queryToCombinations = query.slice(i);
      const allCombinations = findAllCombinations(queryToCombinations);
      const words = this.checkWordsAgainstDictionary(allCombinations);
      for (const word of words) {
        result.push(query.slice(0, i) + word + query.slice(i + word.length));
      }
    }
    return result;
  }

  numberToWords(number) {
    const query = removeNonAlphaNumeric(number);
    this._isValidQuery(query);
    let result = '';
    let lastWord = '';
    for (let i = 2; i < query.length; i++) {
      const queryToCombinations = query.slice(i);
      const allCombinations = findAllCombinations(queryToCombinations);
      const words = this.checkWordsAgainstDictionary(allCombinations);
      for (const word of words) {
        const ans = query.slice(0, i) + word + query.slice(i + word.length);
        if (word.length >= lastWord.length) {
          result = ans;
          lastWord = word;
        }
      }
    }
    return result;
  }

  static wordsToNumber(queryString) {
    return stringToNumber(queryString);
  }
}
